### PR TITLE
Socket Connection Handler Updates

### PR DESCRIPTION
**Pull request by nikonratm**
Part of addressing #88

@nzapponi Please take a look at this. I think there were two separate issues, one was rate-limiting but the other was left un-addressed which is that the current code ends up creating multiple listening objects resulting in logs like:
```
[4/22/2020, 2:31:46 PM] [Home Alarm] Received new event from alarm: ALARM_DISARM
[4/22/2020, 2:31:46 PM] [Home Alarm] Received new event from alarm: ALARM_DISARM
[4/22/2020, 2:31:46 PM] [Home Alarm] Received new event from alarm: ALARM_DISARM
[4/22/2020, 2:31:46 PM] [Home Alarm] Received new event from alarm: ALARM_DISARM
[4/22/2020, 2:31:46 PM] [Home Alarm] Received new event from alarm: ALARM_DISARM
[4/22/2020, 2:31:46 PM] [Home Alarm] Received new event from alarm: ALARM_DISARM
[4/22/2020, 2:31:46 PM] [Home Alarm] Received new event from alarm: ALARM_DISARM
```

I think that is because we were passing `DISCONNECT` events from SS back to the alarm and the camera and [they call `startListening()` again](https://github.com/nzapponi/homebridge-simplisafe3/blob/e7b79f307e1ec1af9d771bba44a08d17644c8144/src/accessories/alarm.js#L205)  but that is actually unnecessary because the [.io protocol defaults to automatically reconnecting](https://socket.io/docs/client-api/#new-Manager-url-options). So as is the plugin would automatically _and manually_ re-connect, resulting in duplicate "listeners". In this PR I changed it to only call `startListening()` when re-connecting ultimately fails ([the `reconnect_failed` event is supposed to fire after the total attempts is tried](https://socket.io/docs/client-api/#Event-‘reconnect-failed’) but the default is infinity attempts). With these changes I now see logs that make more sense, like:
```
[4/23/2020, 7:52:55 AM] [Home Alarm] Living Room camera real time events disconnected.
[4/23/2020, 7:52:55 AM] [Home Alarm] Front Door camera real time events disconnected.
[4/23/2020, 7:52:57 AM] [Home Alarm] Received new event from alarm: RECONNECT
[4/23/2020, 7:52:57 AM] [Home Alarm] Living Room camera real time events re-connected.
[4/23/2020, 7:52:57 AM] [Home Alarm] Front Door camera real time events re-connected.
```

The only thing Im a little unsure of are the changes I made to simplisafe.js 764-774. I removed the built-in handlers for `'error'` and `'disconnect'` because I was thinking we dont need to destroy the socket connection on disconnects (Im less certain about removing the error handler). Incidentally I was wondering if those handlers should use your `unsubscribeFromEvents()` function.

Let me know your thoughts—cheers!

### CODE REVIEW

**Comment by nzapponi** [comment]
Thanks for pulling this together.
I mostly agree with your proposed changes - the one I'm a little unsure about is the one you mention in simplisafe.js... I'm worried that not handling the error means that at some point we might end up without a listener altogether.

So basically what you're saying is that if the socket were to drop, you'd get a `DISCONNECT` and then hopefully a `RECONNECT`. But, if there were to be a `RECONNECT_FAILED`, then simplisafe.js would call `unsubscribeFromSensors()` and destroy the socket, at the same time send a callback to alarm.js and simplicam.js, which would then trigger `startListening()` again. Is this right? Have you tested this flow?

Thanks!

File: src/accessories/alarm.js [hunk 1]
```diff
@@ -201,10 +200,17 @@ class SS3Alarm {
...
                             break;
                         case EVENT_TYPES.DISCONNECT:
-                            this.log('Real time events disconnected.');
+                            this.log('Alarm real time events disconnected.');
+                            break;
+                        case EVENT_TYPES.RECONNECT:
+                            this.log('Alarm real time events re-connected.');
+                            break;
+                        case EVENT_TYPES.RECONNECT_FAILED:
+                            this.log(this.name + ' camera real time events re-connect failed. Restarting');
```
Comment (nzapponi) [inline src/accessories/alarm.js:209]: Change this log to `this.log(this.name + ' alarm real time events re-connect failed. Restarting');`

**Comment by nikonratm** [comment]
Yea thats the basic idea. The flow seems to work but its not perfect yet. There seems to be some time(s) when it never re-connects, trying to see if I can figure out how to 'detect' those. Alternatively I considered leaving the old logic and just making sure to destroy any old listeners, but this makes more sense dont you think? I'll keep testing and see if I can nail it down completely, your thoughts / testing obviously appreciated.

**Comment by nikonratm** [comment]
I updated with some new logic, basically on `error` and on `reconnect_failed` it closes & destroys the socket and then lets the alarm / camera re-create its listener. I am still testing this but seems more reliable.

Can I ask a question though, why did you decide to ever get rid of the `socket` object? As in if we just made sure to catch any errors / permanent disconnects and then [re-start the connection](https://socket.io/docs/client-api/#Event-‘disconnect’) we would never have to call `startListening` again because we could ensure the same socket object is always available. Im not super familiar with the socket.io API so perhaps you did that because you found it more reliable? Just curious.

**Comment by nikonratm** [comment]
OK! So still testing but just documenting latest changes. Besides some re-factoring I've added a separate `switch` block to all `startListening` functions that is exclusively for socket events (because for the alarm which gets instantiated first `this.service` is null at that point but we still want it to handle socket events). The other thing is I've consolidated the socket lost events into `CONNECTION_LOST` (e.g. for `error`, `reconnect_failed`, etc). Each of those handlers uses the same `unsubscribeFromEvents()` which also now explicitly includes `socket.off()` to un-hook any listeners. Then the camera / alarm / whatever is cleared to setup its listeners again.

Again, still testing...

**Comment by nikonratm** [comment]
I think Ive got this flow working pretty well, sample output is below which kinda demonstrates. This is when I disconnect my entire internet. I would like to test more real-world situations but Im not sure if its because of these changes or just random luck but I have had so few socket disconnects and rate limit errors its tough to test! Did you mention at some point a method to test those?

```
[4/25/2020, 3:57:58 PM] [Home Alarm] Socket disconnect reason: ping timeout
[4/25/2020, 3:57:58 PM] [Home Alarm] Alarm real time events disconnected.
[4/25/2020, 3:57:58 PM] [Home Alarm] Living Room camera real time events disconnected.
[4/25/2020, 3:57:58 PM] [Home Alarm] Front Door camera real time events disconnected.
[4/25/2020, 3:57:59 PM] [Home Alarm] Socket reconnect_attempt 1
[4/25/2020, 3:57:59 PM] [Home Alarm] Socket connect_error TransportError: websocket error
[4/25/2020, 3:57:59 PM] [Home Alarm] Alarm real time events connection lost. Attempting to reconnect...
[4/25/2020, 3:57:59 PM] [Home Alarm] Living Room camera real time events connection lost. Attempting to reconnect...
[4/25/2020, 3:57:59 PM] [Home Alarm] Front Door camera real time events connection lost. Attempting to reconnect...
[4/25/2020, 3:58:53 PM] [Home Alarm] Socket connect_error TransportError: websocket error
[4/25/2020, 3:58:53 PM] [Home Alarm] Alarm real time events connection lost. Attempting to reconnect...
[4/25/2020, 3:58:53 PM] [Home Alarm] Living Room camera real time events connection lost. Attempting to reconnect...
[4/25/2020, 3:58:53 PM] [Home Alarm] Front Door camera real time events connection lost. Attempting to reconnect...
...
[4/25/2020, 3:59:56 PM] [Home Alarm] Socket connect
[4/25/2020, 3:59:56 PM] [Home Alarm] Alarm now listening for real time events.
[4/25/2020, 3:59:56 PM] [Home Alarm] Living Room camera now listening for real time events.
[4/25/2020, 3:59:56 PM] [Home Alarm] Front Door camera now listening for real time events.
```

**Comment by nzapponi** [comment]
Cool thanks for all of this @nikonratm !
I honestly don't remember why I set it up the way I did, so if this flow works better, I'm all for it.

I can help test it over the next week or so. Have you considered running Homebridge on your machine and using a proxy such as Charles Proxy to control the network state and fake responses from SimpliSafe?

**Comment by nikonratm** [comment]
Charles is awesome, thanks—much easier!

So far so good in debugging. From digging into this I _think_ that `disconnect`, `connect_timeout` and `connect_error` will all still automatically re-connect so they dont need to re-create listeners. `error` and `reconnect_failed` seem to be permanent so they are the only two that destroy the socket and pass along `CONNECTION_LOST` to start things up again.

Ok let me know if you see any issues, I'll keep it running.

BTW, unrelated, do you use ESLint / do you want to publish your .eslintrc ?

**Comment by nzapponi** [comment]
Awesome, thanks for the great work, works really well! I'll release it and let's keep an eye out for issues.

**Pull request merged**